lib/spack/spack/main.py

```diff
@@ -0,0 +1,464 @@
+##############################################################################
+# Copyright (c) 2013-2017, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+"""This is the implementation of the Spack command line executable.
+
+In a normal Spack installation, this is invoked from the bin/spack script
+after the system path is set up.
+"""
+from __future__ import print_function
+
+import sys
+import os
+import inspect
+from argparse import _ArgumentGroup, ArgumentParser, RawTextHelpFormatter
+import pstats
+
+import llnl.util.tty as tty
+from llnl.util.tty.color import *
+
+import spack
+import spack.cmd
+from spack.error import SpackError
+
+
+# names of profile statistics
+stat_names = pstats.Stats.sort_arg_dict_default
+
+# help levels in order of detail (i.e., number of commands shown)
+levels = ['short', 'long']
+
+# intro text for help at different levels
+intro_by_level = {
+    'short': 'These are common spack commands:',
+    'long':  'Complete list of commands:',
+}
+
+# control top-level spack options shown in basic vs. advanced help
+options_by_level = {
+    'short': 'hkV',
+    'long': 'all'
+}
+
+# Longer text for each section, to show in help
+section_descriptions = {
+    'admin':       'administration',
+    'basic':       'basic usage',
+    'build':       'build pipeline (in order of execution)',
+    'config':      'configuration',
+    'developer':   'developer',
+    'environment': 'environment',
+    'extensions':  'extensions',
+    'help':        'further help',
+    'packaging':   'packaging',
+    'system':      'system',
```

Comment:
Cool.  We can revisit this later if we want.  But `argparse` is really nasty and I'm trying to avoid modifying its internals.  The `argparse` package we have in there is a backport to support 2.6; nothing custom.  I'd rather keep it that way.